lib/services/local/access_graph.go

```diff
@@ -0,0 +1,178 @@
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	privateKeysSvc, err := generic.NewServiceWrapper(
+		backend,
+		types.KindAccessGraphSecretPrivateKey,
+		privateKeysPrefix,
+		services.MarshalAccessGraphPrivateKey,
+		services.UnmarshalAccessGraphPrivateKey,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &AccessGraphSecretsService{
+		authorizedKeysSvc: authorizedKeysSvc,
+		privateKeysSvc:    privateKeysSvc,
+	}, nil
+}
+
+// ListAllAuthorizedKeys lists all authorized keys stored in the backend.
+func (k *AccessGraphSecretsService) ListAllAuthorizedKeys(ctx context.Context, pageSize int, pageToken string) ([]*accessgraphsecretspb.AuthorizedKey, string, error) {
+	out, next, err := k.authorizedKeysSvc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// ListAuthorizedKeysForServer lists all authorized keys for a given hostID.
+func (k *AccessGraphSecretsService) ListAuthorizedKeysForServer(ctx context.Context, hostID string, pageSize int, pageToken string) ([]*accessgraphsecretspb.AuthorizedKey, string, error) {
+	if hostID == "" {
+		return nil, "", trace.BadParameter("server name is required")
+	}
+	svc := k.authorizedKeysSvc.WithPrefix(hostID)
+	out, next, err := svc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// UpsertAuthorizedKey upserts a new authorized key.
+func (k *AccessGraphSecretsService) UpsertAuthorizedKey(ctx context.Context, in *accessgraphsecretspb.AuthorizedKey) (*accessgraphsecretspb.AuthorizedKey, error) {
+	svc := k.authorizedKeysSvc.WithPrefix(in.Spec.HostId)
+	out, err := svc.UpsertResource(ctx, in)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return out, nil
+}
+
+// DeleteAuthorizedKey deletes a specific authorized key.
+func (k *AccessGraphSecretsService) DeleteAuthorizedKey(ctx context.Context, hostID, name string) error {
+	svc := k.authorizedKeysSvc.WithPrefix(hostID)
+	return trace.Wrap(svc.DeleteResource(ctx, name))
+}
+
+// DeleteAllAuthorizedKeys deletes all authorized keys.
+func (k *AccessGraphSecretsService) DeleteAllAuthorizedKeys(ctx context.Context) error {
+	return trace.Wrap(k.authorizedKeysSvc.DeleteAllResources(ctx))
+}
+
+// ListAllPrivateKeys lists all private keys stored in the backend.
+func (k *AccessGraphSecretsService) ListAllPrivateKeys(ctx context.Context, pageSize int, pageToken string) ([]*accessgraphsecretspb.PrivateKey, string, error) {
+	out, next, err := k.privateKeysSvc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// ListPrivateKeysForDevice lists all private keys for a given deviceID.
+func (k *AccessGraphSecretsService) ListPrivateKeysForDevice(ctx context.Context, deviceID string, pageSize int, pageToken string) ([]*accessgraphsecretspb.PrivateKey, string, error) {
```

Comment:
At first glance this look scary that we are listing PrivateKeys. Can we just rework this to `ListSSHKeysDetails` ?

Reply:
These are not private keys. We call them private keys because they represent a private key in the host machine but they actually just store the fingerprint of the public key derived from a private key. 

There is no private key material leaving the user's laptop so this is 100% safe. 
besides that, there won't be a public api to export these findings to the outside.

Reply:
I agree with Marek. If this API doesn't list private keys, I'd not call it "ListPrivateKeys".